src/analysis/processing/qgsalgorithmextractbyexpression.cpp

```diff
@@ -156,6 +156,14 @@ QVariantMap QgsExtractByExpressionAlgorithm::processAlgorithm( const QVariantMap
     }
   }
 
+  if ( matchingSink && !matchingSink->flushBuffer() )
+  {
+    throw QgsProcessingException( matchingSink->lastError() );
+  }
```

Comment:
If we have to do that in all algorithms, perhaps have a flushBufferAndThrowProcessingExceptionIfError() method (or if not appropriate in the feature sink class, in some utility class), to be able to do:
```suggestion
  if ( matchingSink )
  {
    matchingSink->flushBufferAndThrowProcessingExceptionIfError();
  }
```

Other idea: have parameterAsSink() return a feature sink that does this if (!flushBufer()) ) throw QgsProcessingException(...) in its destructor.

Reply:
@rouault 

> throw QgsProcessingException(...) in its destructor.

That was my original plan, but throwing exceptions is explicitly illegal in destructors. 

Other thoughts I had:

- use the destructor to mark a "throw exception" member in the processing context, and throw the actual exception from elsewhere (pretty messy)
- add some "finalize" method on the sink which flushes the buffer and raises the exception. Algorithms would still need to explicitly call this, but it's a little nicer then flushBuffer calls
- for python we could maybe hack in some __del__ magic in the sink which raises the exception when the python wrapper is deleted (although I'm unsure if the python wrapper is always deleted before the C++ object....)
- ???

Reply:
> add some "finalize" method on the sink which flushes the buffer and raises the exception. Algorithms would still need to explicitly call this, but it's a little nicer then flushBuffer calls

This looks like the appropriate way of doing it. fstream [example](https://stackoverflow.com/questions/130117/if-you-shouldnt-throw-exceptions-in-a-destructor-how-do-you-handle-errors-in-i) looks like very much what we have here, meaning close/finalize would trigger an exception that can be dealt by calling code, but its destructor would not.

Reply:
Thanks @troopa81 , that's a valuable link! 

I saw with interest this later reply : https://stackoverflow.com/a/41429901 which looks potentially like a safe way to throw in the destructor. What do you think about that approach? (Also cc @rouault )

Reply:
I have mixed feelings. On one hand it looks like complicated, maybe difficult it debug, a workaround of something set illegal in the language for good reason. On the other hand it solves the issue for all the algorithms, even the python plugin ones, without modifying the algorithms.

So, why not.

Reply:
@troopa81 ok, that's a no go -- we can't mark the destructor as noexcept(false) because it overrides the base class destructor and we get ` error: looser exception specification on overriding virtual function ` -- and we can't change the base class destructor unless we change ALL the featuresink subclass destructors to noexcept... and we CAN'T do that because some of them also inherit QObject. :upside_down_face: 

I'll go with explicit finalize calls instead.